### ISSUE TITLE
Crest requires lots of layers

### ISSUE THREAD

**Issue by gamemachine** [huwb/crest-oceanrender#61]
Number of layers required is a bit much.  The demo requires 4 (Terrain is not mentioned that I could find in the docs).  This actually used up all remaining layers in my project.   Layers tend to be at a premium in large games.

**Comment by gamemachine** [comment]
Saw the ocean needs a layer also.  That's 5 layers now.

**Comment by huwb** [comment]
Yeah, this comes up now and then, theres another team in a similar situation and I'm becoming increasingly aware of the squeeze. It's a shame that the number of layers is so small, and that physics and rendering share layers. But thats the constraints we have i guess!

It's there for flexibility so you can assign whatever you like into the various data like water shape, foam, dynamic sim, etc. If you're not using any of these you can not add the relevant layer (i've not done a huge amount testing in this direction so it might need one or two things unpicked but it should be easy to make it work with only the layers you want/need).

However I get the point and this kind of functionality could perhaps be achieved using tags. Scriptable render pipelines may change this up as well, which is something i plan to do research into soon. I'll leave this open and post back when i have any thoughts or progress, or perhaps someone else looks at it in the meantime.

**Comment by huwb** [comment]
Yeah im thinking along the same lines.. In my past experience I haven't loved doing this kind of thing because theres the state to manage/update when new objects get created, and when stuff gets destroyed you dont want to hang on to references and prevent stuff getting GC'd. Or maybe im making a mountain out of a mole hill. I'll put it on my list to experiment with this stuff soon and find a nice pattern.

**Comment by gamemachine** [comment]
I think it would involve more work then appears at first glance.  It also looks like it wouldn't be too much work to just add work arounds where needed per project until a better overall solution hits.  Big picture I don't see this as a show stopper for real teams.

**Comment by huwb** [comment]
There is some evidence from only THIRTEEN YEARS ago that just doing a FindObjectsWithTag may not be as bad as it sounds, in practice (which would remove the state that needs to be maintained). It seems to be efficient provided you don't use tags for too many GOs, which he suggests people typically dont. Do you use tags much in your projects?

https://forum.unity.com/threads/findgameobjectswithtag-efficiency.3612/

**Comment by gamemachine** [comment]
I mostly see tags used by third party assets.  They do seem like an ok integration point assuming that in context they won't be a performance issue.  And they aren't a limited resource which is a big plus.

One potential downside is what stuff created at runtime.  And I say potential just because I have no idea of the possible issues there not having used tags in a long time.

**Comment by huwb** [comment]
In preparation for a new release i've just gone through the initial setup thing with the layers and its clearly demanding layers that are not necessarily needed (like the dynamic wave sim layer even when the sim is disabled), so i'll look at addressing this.

**Comment by huwb** [comment]
Ok i'm being much more sensitive to the layer situation now on latest/master. It only throws errors if the layer is missing and *actually required* in the scene.

I also removed the Terrain layer which renders the sea floor height - I know of at least one team that is just rendering Default layer geometry into this, and this works for the example scene, so I'm going with this too.

That means that the main scene in the example content only requires **2 layers** - for the animated waves and for the foam (the ocean uses the Water layer but this is built-in and the right layer for the job).

While I think it would be possible to explore some stuff we discussed above (switching to tags for instance), I don't currently have plans to do so and don't see a very clear/easy path to doing so, and will therefore close this issue for now. If you have thoughts, let me know.

( @holdingjason @dizzy2003 )
